Router remove directive interceptor

petebacondarwin opened the pull request:
This PR sits on top of #7126 which should land soon.

petebacondarwin committed:
refactor(angular_1_router): remove directiveIntrospector

The directiveIntrospector was a bit of a hack to allow the router to
read the `$routeConfig` annocation and `$routerCanActivate` hook from
directives when they were registered.

It turns out that if we put these properties on the component controller's
constructor function (i.e. as static class methods) then we can simply
use the `$injector` to access it as required.

Currently, people put the properties directly on their component definition
objects. In Angular 1.5.1, we will copy these properties onto the controller
constructor to maintain a simple migration path. But going forward it may be
better to encourage people to add the properties directly to the controller
constructor.

petebacondarwin committed:
test(angular_1_router): apply annotations to controller constructors

Until Angular 1.5.1 is released, the `$routeConfig` and `$routerCanActivate`
annotations for components must live on the controller constructor.

In Angular 1.5.1, it will automatically copy these annotations across from
the component definition file.

btford commented:
Looks good, but two tests are failing (missing a mocked out `$rootScope` it looks like).

petebacondarwin committed:
test(angular_1_router): fix router_spec tests

These tests were registering new components after the application had
been bootstrapped, which is not a valid use case for synchronous routes
in Angular 1.

In particular it was registering the "root" component, which caused the
`$rootRouter` to blow up, when it was instantiated, pointing to a root
component that did not yet exist.

mary-poppins commented:
User @martine does not have PR merging privileges.

martine commented:
g3sync says this is ok, but I currently lack the admin rights to merge it.  Working on it.

mary-poppins commented:
Merging PR #7319 on behalf of @martine to branch [presubmit-martine-pr-7319](https://github.com/angular/angular/tree/presubmit-martine-pr-7319).
